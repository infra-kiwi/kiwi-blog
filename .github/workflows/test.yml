name: Run tests

on:
  pull_request:
    paths-ignore:
      - '**/*.md'

jobs:
  find-changes:
    name: Find changed directories
    runs-on: ubuntu-22.04
    outputs:
      matrix: ${{ steps.find-changes.outputs.matrix }}
      matrix-empty: ${{ steps.find-changes.outputs.matrix-empty }}
    steps:
      - uses: actions/checkout@v4
        with:
          fetch-depth: 0
      - id: find-changes
        uses: cmaster11/gha/action-get-changed-dirs@main
        with:
          base-sha: ${{ github.event.pull_request.base.sha }}
          regex: '^(\d+-|boilerplate-cdk)'

  test:
    needs: find-changes
    if: needs.find-changes.outputs.matrix-empty == 'false'
    name: Run tests
    runs-on: ubuntu-22.04
    strategy:
      matrix: ${{ fromJSON(needs.find-changes.outputs.matrix) }}
      fail-fast: false
    steps:
      - uses: actions/checkout@v4
        with:
          fetch-depth: 0
      - uses: actions/setup-node@v4
        with:
          node-version-file: ${{ matrix.directory }}/package.json
      - name: Setup deps
        working-directory: ${{ matrix.directory }}
        run: npm ci
      - name: Lint
        working-directory: ${{ matrix.directory }}
        run: npm run lint
      - name: Test
        working-directory: ${{ matrix.directory }}
        run: npm run test:ci

  dependabot-fixer:
    name: Dependabot fixer
    needs: test
    if: >
      failure() && 
      github.event.pull_request.user.login == 'dependabot[bot]'
    runs-on: ubuntu-22.04
    steps:
      - name: Dependabot metadata
        id: dependabot-metadata
        uses: dependabot/fetch-metadata@v1
      - uses: actions/checkout@v4
        with:
          ref: ${{ github.head_ref }}
      - shell: pwsh
        run: |
          ./hack/fix-all.ps1
      - run: |
          if git diff --quiet -- */test/__snapshots__/*.snap && git diff --cached --quiet -- */test/__snapshots__/*.snap; then
            echo "No changes"
            exit 0
          fi
          
          branch="${GITHUB_HEAD_REF:-${GITHUB_REF#refs/heads/}}"
          
          # https://blog.infra.kiwi/pushing-new-commits-on-github-actions-with-a-personal-access-token-pat-8c9240fbc8d2
          git config --unset "http.https://github.com/.extraheader"
          git remote set-url origin "https://x-access-token:${DEPENDABOT_PAT}@github.com/${GITHUB_REPOSITORY}.git/"
          
          git add */test/__snapshots__/*.snap
          git commit -m "Autofix dependabot [dependabot skip]"
          git push origin "HEAD:$branch"
        env:
          DEPENDABOT_PAT: ${{ secrets.DEPENDABOT_PAT }}

      - uses: actions/github-script@v7
        with:
          script: |
            github.rest.issues.createComment({
              issue_number: context.issue.number,
              owner: context.repo.owner,
              repo: context.repo.repo,
              body: '@dependabot rebase'
            })
          github-token: ${{ secrets.DEPENDABOT_PAT }}