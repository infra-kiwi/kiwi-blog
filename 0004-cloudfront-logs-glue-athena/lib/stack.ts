import * as cdk from 'aws-cdk-lib';
import { aws_cloudfront, aws_s3, CfnOutput, RemovalPolicy, Stack } from 'aws-cdk-lib';
import { Construct } from 'constructs';
import { HttpOrigin } from 'aws-cdk-lib/aws-cloudfront-origins';
import * as aws_glue_alpha from '@aws-cdk/aws-glue-alpha';

export class CloudFrontStack extends cdk.Stack {
  constructor(scope: Construct, id: string, props?: cdk.StackProps) {
    super(scope, id, props);

    // Create a dedicated S3 bucket to store CloudFront logs
    const logBucket = new aws_s3.Bucket(this, 'CloudFrontLogs', {
      // ACL, required by CloudFront
      objectOwnership: aws_s3.ObjectOwnership.BUCKET_OWNER_PREFERRED,
      // Make sure we don't keep unnecessary data after we delete this stack
      removalPolicy: RemovalPolicy.DESTROY,
      autoDeleteObjects: true
    });

    const distribution = new aws_cloudfront.Distribution(this, 'Distribution', {
      // HTTPS (because we should ALWAYS use HTTPS!)
      sslSupportMethod: aws_cloudfront.SSLMethod.SNI,

      // Enable logging to the S3 bucket
      enableLogging: true,
      logBucket: logBucket,

      // Use the wonderful https://reqres.in/ as a destination for our calls
      defaultBehavior: {
        origin: new HttpOrigin('reqres.in')
      }
    });
    // This output will contain the domain name of the distribution,
    // e.g. xxxxxxx.cloudfront.net
    new CfnOutput(this, 'DistributionDomainName', {
      value: distribution.domainName
    });

    // Create a database where to store the table
    const database = new aws_glue_alpha.Database(this, 'GlueDatabase', {
      // Give the database a friendly name (lower-case only!)
      databaseName: Stack.of(this).stackName + '-database'
    });
    new CfnOutput(this, 'CloudFrontLogsDatabaseName', {
      value: database.databaseName
    });

    // Create the table that maps each log line
    const table = new aws_glue_alpha.S3Table(this, 'CloudFrontLogsTable', {
      // Give the table a friendly name (lower-case only!)
      tableName: Stack.of(this).stackName + '-cloudfrontlogstable',

      // The source bucket for this Glue table
      bucket: logBucket,
      database: database,

      // Full SQL reference: https://docs.aws.amazon.com/athena/latest/ug/cloudfront-logs.html
      columns: [
        { name: 'date', type: aws_glue_alpha.Schema.DATE },
        { name: 'time', type: aws_glue_alpha.Schema.STRING },
        { name: 'x_edge_location', type: aws_glue_alpha.Schema.STRING },
        { name: 'sc_bytes', type: aws_glue_alpha.Schema.BIG_INT },
        { name: 'c_ip', type: aws_glue_alpha.Schema.STRING },
        { name: 'cs_method', type: aws_glue_alpha.Schema.STRING },
        { name: 'cs_host', type: aws_glue_alpha.Schema.STRING },
        { name: 'cs_uri_stem', type: aws_glue_alpha.Schema.STRING },
        { name: 'sc_status', type: aws_glue_alpha.Schema.INTEGER },
        { name: 'cs_referrer', type: aws_glue_alpha.Schema.STRING },
        { name: 'cs_user_agent', type: aws_glue_alpha.Schema.STRING },
        { name: 'cs_uri_query', type: aws_glue_alpha.Schema.STRING },
        { name: 'cs_cookie', type: aws_glue_alpha.Schema.STRING },
        { name: 'x_edge_result_type', type: aws_glue_alpha.Schema.STRING },
        { name: 'x_edge_request_id', type: aws_glue_alpha.Schema.STRING },
        { name: 'x_host_header', type: aws_glue_alpha.Schema.STRING },
        { name: 'cs_protocol', type: aws_glue_alpha.Schema.STRING },
        { name: 'cs_bytes', type: aws_glue_alpha.Schema.BIG_INT },
        { name: 'time_taken', type: aws_glue_alpha.Schema.FLOAT },
        { name: 'x_forwarded_for', type: aws_glue_alpha.Schema.STRING },
        { name: 'ssl_protocol', type: aws_glue_alpha.Schema.STRING },
        { name: 'ssl_cipher', type: aws_glue_alpha.Schema.STRING },
        { name: 'x_edge_response_result_type', type: aws_glue_alpha.Schema.STRING },
        { name: 'cs_protocol_version', type: aws_glue_alpha.Schema.STRING },
        { name: 'fle_status', type: aws_glue_alpha.Schema.STRING },
        { name: 'fle_encrypted_fields', type: aws_glue_alpha.Schema.INTEGER },
        { name: 'c_port', type: aws_glue_alpha.Schema.INTEGER },
        { name: 'time_to_first_byte', type: aws_glue_alpha.Schema.FLOAT },
        { name: 'x_edge_detailed_result_type', type: aws_glue_alpha.Schema.STRING },
        { name: 'sc_content_type', type: aws_glue_alpha.Schema.STRING },
        { name: 'sc_content_len', type: aws_glue_alpha.Schema.BIG_INT },
        { name: 'sc_range_start', type: aws_glue_alpha.Schema.BIG_INT },
        { name: 'sc_range_end', type: aws_glue_alpha.Schema.BIG_INT }
      ],

      // Define the right settings to parse CloudFront logs
      dataFormat: aws_glue_alpha.DataFormat.TSV,
      parameters: {
        // THESE PARAMETERS ARE VERY IMPORTANT
        'skip.header.line.count': '2',
        'serialization.format': '\t',
        'field.delim': '\t'
      }
    });
    new CfnOutput(this, 'CloudFrontLogsTableName', {
      value: table.tableName
    });
  }
}
