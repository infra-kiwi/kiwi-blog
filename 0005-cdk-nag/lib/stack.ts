/*
 * Copyright (c) 2024. Alberto Marchetti - https://www.linkedin.com/in/albertomarchetti/
 */

import * as cdk from 'aws-cdk-lib';
import { aws_kms, aws_sns, CfnOutput } from 'aws-cdk-lib';
import { Construct } from 'constructs';

export class KiwiStack extends cdk.Stack {
  readonly topic: aws_sns.Topic;

  constructor(scope: Construct, id: string, props?: cdk.StackProps) {
    super(scope, id, props);

    // Use the default AWS-provided SNS KMS key
    const awsSNSKey = aws_kms.Alias.fromAliasName(
      this,
      'AWSManagedSNSKey',
      'alias/aws/sns'
    );

    this.topic = new aws_sns.Topic(this, 'MyTopic', {
      masterKey: awsSNSKey
    });
    new CfnOutput(this, 'MyTopicArn', {
      value: this.topic.topicArn
    });
  }
}
