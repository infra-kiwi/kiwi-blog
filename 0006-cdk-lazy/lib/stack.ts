import * as cdk from 'aws-cdk-lib';
import { aws_sns, Stack } from 'aws-cdk-lib';
import { Construct } from 'constructs';
import { MultiLineParameter } from './multi-line-parameter';

export class KiwiStack extends cdk.Stack {
  readonly topic: aws_sns.Topic;

  constructor(scope: Construct, id: string, props?: cdk.StackProps) {
    super(scope, id, props);

    const myParameter = new MultiLineParameter(this, 'MyParameter', {
      parameterName: Stack.of(this).stackName + '-my-parameter'
    });

    // Populate the lines of our multi-line parameter
    myParameter.addLine("It's not DNS");
    myParameter.addLine("There's no way it's DNS");
    myParameter.addLine('It was DNS');

    // Add a dynamic line that depends on another CDK resource's value
    const aTopicAboutDNS = new aws_sns.Topic(this, 'ATopicAboutDNS');
    myParameter.addLine(
      `Here is a topic about DNS: ${aTopicAboutDNS.topicArn}`
    );
    this.topic = aTopicAboutDNS;
  }
}
