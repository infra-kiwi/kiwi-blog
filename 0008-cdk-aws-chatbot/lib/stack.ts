/*
 * Copyright (c) 2024. Alberto Marchetti - https://www.linkedin.com/in/albertomarchetti/
 */

import * as cdk from 'aws-cdk-lib';
import { Construct } from 'constructs';
import { ChatBotSlackChannel } from './constructs/chatbot-slack-channel';
import { PostToSlack } from './constructs/post-to-slack';
import { Config } from './util/config';

export type ChatBotStackProps = cdk.StackProps & {
  config: Config;
};

export class ChatBotStack extends cdk.Stack {
  constructor(scope: Construct, id: string, props: ChatBotStackProps) {
    super(scope, id, props);

    const chatBotSlackChannels: Record<string, ChatBotSlackChannel> = {};

    // Create all the possible ChatBot channel configurations. To do so,
    // extract all available Slack workspace/channel pairs, deduplicate them,
    // and create a ChatBotSlackChannel for each
    const slackChannelEntries = Array.from(
      new Set(
        Object.entries(props.config.notificationChannels).map(([, v]) =>
          JSON.stringify([v.slackWorkspaceId, v.slackChannelId])
        )
      )
    ).map((e) => JSON.parse(e)) as [string, string][];

    for (const entry of slackChannelEntries) {
      const [slackWorkspaceId, slackChannelId] = entry;

      // Configure the channel
      const channel = new ChatBotSlackChannel(
        this,
        `ChatBotSlackChannel-${slackWorkspaceId}-${slackChannelId}`,
        {
          slackWorkspaceId,
          slackChannelId
        }
      );
      chatBotSlackChannels[JSON.stringify(entry)] = channel;
    }

    // Create all the PostToSlack instances, which will accept our custom payloads
    // and translate them into ChatBot-formatted ones
    for (const notificationChannelName in props.config.notificationChannels) {
      const config = props.config.notificationChannels[notificationChannelName];

      // Use the previously-generated Slack channel configuration
      const chatBotSlackChannel =
        chatBotSlackChannels[
          JSON.stringify([config.slackWorkspaceId, config.slackChannelId])
        ];

      new PostToSlack(this, `SlackLambda-${notificationChannelName}`, {
        chatbotSNSTopicArn: chatBotSlackChannel.topicArn,
        notificationChannelName
      });
    }
  }
}
