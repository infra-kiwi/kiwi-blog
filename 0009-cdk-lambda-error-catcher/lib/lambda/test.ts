export const handler = ({ fail }: { fail?: boolean | 'log' }) => {
  if (fail === true) {
    throw new Error('This error was thrown!');
  }

  if (fail === 'log') {
    console.error('This is an error logged via console.error');
  } else {
    console.log('All good!');
  }
};
