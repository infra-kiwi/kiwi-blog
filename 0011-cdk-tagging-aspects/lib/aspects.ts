import * as cdk from 'aws-cdk-lib';
import { Annotations, aws_sns, TagManager } from 'aws-cdk-lib';
import { IConstruct } from 'constructs';

// Defines an aspect that will visit every resource of the Stack and
// assign a tag value based on the path of such resource.
//
// E.g. If the main Stack ID is "Test", an SNS resource
// could have the path Test/MyTopic/Resource.
export class PathTaggerAspect implements cdk.IAspect {
  visit(node: IConstruct) {
    new cdk.Tag('aws-cdk-path', node.node.path).visit(node);
  }
}

// Defines an aspect that will print info about every single node
// visited by the Aspect
export class DebugAspect implements cdk.IAspect {
  visit(node: IConstruct) {
    const className = node.constructor.name;
    const tags = TagManager.of(node)?.tagValues();

    console.log(`Visiting ${className}`, {
      id: node.node.id,
      tags
    });
  }
}

// Defines an aspect that make sure every SNS topic has a specific tag
// defined, otherwise it will add an annotation to prevent any successful
// synth of the Stack.
export class ValidateSNSTopicAspect implements cdk.IAspect {
  visit(node: IConstruct) {
    if (!(node instanceof aws_sns.CfnTopic)) {
      return;
    }

    if (!('MyRequiredTag' in (TagManager.of(node)?.tagValues() ?? {}))) {
      Annotations.of(node).addError(
        `Missing required tag MyRequiredTag in CfnTopic ${node.logicalId}!`
      );
    }
  }
}

export async function addAspects(
  scope: IConstruct,
  props?: {
    debug?: boolean;
    forcedTags?: Record<string, string>;
  }
) {
  const tags = cdk.Tags.of(scope);
  const aspects = cdk.Aspects.of(scope);

  // Add every forced tag to the Stack
  for (const tag in props?.forcedTags) {
    tags.add(tag, props.forcedTags[tag]);
  }

  // Apply the PathTagger to the whole stack
  aspects.add(new PathTaggerAspect());

  if (props?.debug) {
    // Apply the DebugAspect to the stack
    aspects.add(new DebugAspect());
  }

  // Apply the ValidateSNSTopicAspect to the stack
  aspects.add(new ValidateSNSTopicAspect());
}
