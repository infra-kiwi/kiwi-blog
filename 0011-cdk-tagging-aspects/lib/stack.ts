/*
 * Copyright (c) 2024. Alberto Marchetti - https://www.linkedin.com/in/albertomarchetti/
 */

import * as cdk from 'aws-cdk-lib';
import { aws_sns, CfnOutput } from 'aws-cdk-lib';
import { Construct } from 'constructs';

export class KiwiStack extends cdk.Stack {
  readonly topic: aws_sns.Topic;

  constructor(scope: Construct, id: string, props?: cdk.StackProps) {
    super(scope, id, props);

    this.topic = new aws_sns.Topic(this, 'MyTopic');

    // Define an output for debugging purposes
    new CfnOutput(this, 'MyTopicArnOutput', {
      value: this.topic.topicArn
    });
  }
}
