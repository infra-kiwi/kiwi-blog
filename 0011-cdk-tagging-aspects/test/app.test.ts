/*
 * Copyright (c) 2024-2025. Alberto Marchetti - https://www.linkedin.com/in/albertomarchetti/
 */

import { KiwiStack } from '../lib/stack';
import { App, aws_sns, Stack, Tag } from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';
import { describe } from 'node:test';
import { ArtifactMetadataEntryType } from 'aws-cdk-lib/cloud-assembly-schema';
import { addAspects } from '../lib/aspects';

describe('Aspects', () => {
  test('Main stack', async () => {
    const app = new App();

    // Init stack
    const stack = new KiwiStack(app, 'Test');

    await addAspects(app, {
      // Debug which resources we're visiting in the aspects
      debug: true,
      // Add some forced tags to all CDK resources
      forcedTags: {
        MyTag: 'this-is-a-forced-tag',
        MyRequiredTag: 'this-is-a-required-tag'
      }
    });

    // Prepare for assertions
    const template = Template.fromStack(stack);
    expect(template).toMatchSnapshot();

    template.hasResourceProperties('AWS::SNS::Topic', {
      Tags: [
        // Make sure the resource contains the tag introduced
        // via CDK Aspect
        {
          Key: 'aws-cdk-path',
          Value: 'Test/MyTopic/Resource'
        },

        // Make sure the resource contains the required tag
        {
          Key: 'MyRequiredTag',
          Value: 'this-is-a-required-tag'
        },

        // Make sure the resource contains the forced tag
        {
          Key: 'MyTag',
          Value: 'this-is-a-forced-tag'
        }
      ]
    });
  });

  test('ValidateSNSTopicAspect should fail if the required tag is missing', async () => {
    const app = new App();
    const stack = new Stack(app, 'Test');

    await addAspects(app);

    const topic = new aws_sns.Topic(stack, 'MyTopic');

    const template = Template.fromStack(stack);
    expect(template).toMatchSnapshot();

    // Expect an ERROR annotation to exist on the SNS topic created without MyRequiredTag
    expect(
      topic.node.defaultChild!.node.metadata.find(
        (e) => e.type == ArtifactMetadataEntryType.ERROR
      )
    ).not.toBeUndefined();
  });

  test('ValidateSNSTopicAspect should succeed if the required tag is added to the CfnTopic scope via Tag.visit', async () => {
    const app = new App();
    const stack = new Stack(app, 'Test');

    await addAspects(app);

    const topic = new aws_sns.Topic(stack, 'MyTopic');
    new Tag('MyRequiredTag', 'this-is-a-required-tag').visit(
      topic.node.defaultChild!
    );

    const template = Template.fromStack(stack);
    expect(template).toMatchSnapshot();

    // Expect an ERROR annotation to NOT exist on the SNS topic created without MyRequiredTag
    expect(
      topic.node.defaultChild!.node.metadata.find(
        (e) => e.type == ArtifactMetadataEntryType.ERROR
      )
    ).toBeUndefined();
  });
});
