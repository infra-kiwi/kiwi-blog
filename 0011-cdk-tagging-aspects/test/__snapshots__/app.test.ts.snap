// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Main stack 1`] = `
{
  "Outputs": {
    "MyTopicArnOutput": {
      "Value": {
        "Ref": "MyTopic86869434",
      },
    },
  },
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "MyTopic86869434": {
      "Properties": {
        "Tags": [
          {
            "Key": "aws-cdk-path",
            "Value": "Test/MyTopic/Resource",
          },
          {
            "Key": "MyRequiredTag",
            "Value": "this-is-a-required-tag",
          },
          {
            "Key": "MyTag",
            "Value": "this-is-a-forced-tag",
          },
        ],
      },
      "Type": "AWS::SNS::Topic",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;

exports[`ValidateSNSTopicAspect should fail if the required tag is missing 1`] = `
{
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "MyTopic86869434": {
      "Properties": {
        "Tags": [
          {
            "Key": "aws-cdk-path",
            "Value": "Test/MyTopic/Resource",
          },
        ],
      },
      "Type": "AWS::SNS::Topic",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;

exports[`ValidateSNSTopicAspect should succeed if the required tag is added to the CfnTopic scope via Tag.visit 1`] = `
{
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "MyTopic86869434": {
      "Properties": {
        "Tags": [
          {
            "Key": "aws-cdk-path",
            "Value": "Test/MyTopic/Resource",
          },
          {
            "Key": "MyRequiredTag",
            "Value": "this-is-a-required-tag",
          },
        ],
      },
      "Type": "AWS::SNS::Topic",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
